### DISCUSSION TITLE
Does SUPG in INS momentum kernel automatically cover custom source (external force) terms?

### DISCUSSION THREAD

**Discussion by Minjiang-Zhu** [idaholab/moose#26600] [Q&A General] [answered]
As stated, I defined some new source terms in INS momentum equation. I also noticed the SUPG option in INS momentum kernel. I wonder if custom source terms are automatically covered by this SUPG option; otherwise, I should implement SUPG to new source terms by myself.

**Comment by GiudGiud** [comment]
Hello

No if you want SUPG to be consistent with the new source term, you should modify the computation of the momentum equation strong residual in the INSADTauMaterial (see materials/INSADTauMaterial.h). You can see a lot of the terms there, for example Boussinesq or a coupled force strong residual.

Guillaume

**Reply by Minjiang-Zhu** [reply]
Hi, thanks for clarifying. However, I use the INSMomentumTractionFormRZ, which seems not an AD kernel. Should I keep modifying residual in INSADTauMaterial?

**Reply by lindsayad** [reply]
If your body force is described by a function, then you can use the `forcing_func` parameter in your `INSMomentumTractionFormRZ` block. This will be included consistently in the SUPG stabilization

**Reply by lindsayad** [reply]
Just be careful about the sign. `_ffn` is the `forcing_func` in the code
```c++
Real
INSMomentumBase::computeQpResidual()
{
  Real r = 0;

  // viscous term                                                                                                     
  r += computeQpResidualViscousPart();

  // pressure term                                                                                                    
  if (_integrate_p_by_parts)
    r += _grad_test[_i][_qp](_component) * weakPressureTerm();
  else
    r += _test[_i][_qp] * strongPressureTerm()(_component);

  // body force term                                                                                                  
  r += _test[_i][_qp] * (gravityTerm()(_component) - _ffn.value(_t, _q_point[_qp]));
```

**Reply by Minjiang-Zhu** [reply]
Thanks for your support. However, the force term I need requires gradient calculation and variable coupling. Is function method capable of these?

**Reply by GiudGiud** [reply]
Functions can only depend on space, time, and postprocessors.
So it cannot.

**Reply by Minjiang-Zhu** [reply]
I guess I need to modify the corresponding kernels you and @lindsayad mentioned before. Thanks!

**Reply by Minjiang-Zhu** [reply, marked as answer]
I finally managed to fix the issue. The simplest way I found is to modify the body force expressions in INSBase.C.
